[threaded-animation-resolution] Schedule animation updates from RemoteScrollingCoordinatorProxy on the main thread

graouts opened the pull request:
#### e89a58c3716a2cee16351e230a760ed8e8e0358e
<pre>
[threaded-animation-resolution] Schedule animation updates from RemoteScrollingCoordinatorProxy on the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=268282">https://bugs.webkit.org/show_bug.cgi?id=268282</a>
<a href="https://rdar.apple.com/121839192">rdar://121839192</a>

Reviewed by Simon Fraser.

In 273391@main we added support for scheduling animation resolution for macOS via RemoteLayerTreeEventDispatcher.
On iOS, where we don&apos;t use a scrolling thread, we schedule animations through RemoteScrollingCoordinatorProxy
using a dedicated CADisplayLink in a new WKAnimationDisplayLinkHandler. When that display link fires, we resolve
animations on the main thread by calling the new RemoteAcceleratedEffectStack::applyEffectsFromMainThread().

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromMainThread const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(-[WKAnimationDisplayLinkHandler initWithCoordinator:]):
(-[WKAnimationDisplayLinkHandler dealloc]):
(-[WKAnimationDisplayLinkHandler displayLinkFired:]):
(-[WKAnimationDisplayLinkHandler invalidate]):
(-[WKAnimationDisplayLinkHandler schedule]):
(-[WKAnimationDisplayLinkHandler pause]):
(WebKit::RemoteScrollingCoordinatorProxyIOS::~RemoteScrollingCoordinatorProxyIOS):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):

Canonical link: <a href="https://commits.webkit.org/274065@main">https://commits.webkit.org/274065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff0655cf53f77619eed17028dc5a9d817b6a1686

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40269 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33569 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13827 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31930 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38296 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13990 "Found 1 new test failure: media/video-ended-does-not-hold-sleep-assertion.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33028 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12224 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12159 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41528 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34133 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34150 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38045 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36217 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14163 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13134 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4898 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13477 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
```diff
@@ -207,14 +207,39 @@ - (void)didChangeNominalFramesPerSecond
+        RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay();
+
+    if (m_needsDisplayRefreshCallbacksForAnimation)
+        m_webPageProxy->scrollingCoordinatorProxy()->displayDidRefresh(0);
```
The `0` is mysterious.

We do actually synthesize fake displayIDs for iOS (see `generateDisplayIDFromPageID()`). You should use the correct value.

graouts replied:
Fixed with latest iteration of this PR.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -369,15 +374,57 @@
+
+RemoteLayerTreeDrawingAreaProxyIOS& RemoteScrollingCoordinatorProxyIOS::drawingAreaIOS() const
+{
+    auto* drawingArea = dynamicDowncast<RemoteLayerTreeDrawingAreaProxy>(webPageProxy().drawingArea());
```
Why not just add support for `dynamicDowncast<RemoteLayerTreeDrawingAreaProxyIOS>` ?

graouts replied:
Specialized traits for `DrawingArea` use the `DrawingAreaType` enum which does not contain a specific type for the Mac and iOS variants of `RemoteLayerTreeDrawingAreaProxy`. I'm not sure it's worth changing this system for what is only private use within `RemoteScrollingCoordinatorProxyIOS`.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -62,6 +63,10 @@
 {
 }
 
+RemoteScrollingCoordinatorProxyIOS::~RemoteScrollingCoordinatorProxyIOS()
```
`= default`

graouts replied:
Will fix when landing.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -385,15 +390,57 @@
+    auto animatedNodeLayerIDs = std::exchange(m_animatedNodeLayerIDs, { });
+    for (auto animatedNodeLayerID : animatedNodeLayerIDs) {
+        auto* animatedNode = layerTreeHost.nodeForID(animatedNodeLayerID);
+        ASSERT(animatedNode);
```
Debug only assertions when you'll crash on the next line aren't really adding anything.

graouts replied:
Will remove that one and the one two lines below as well.

mattwoodrow committed:
[threaded-animation-resolution] Schedule animation updates from RemoteScrollingCoordinatorProxy on the main thread
https://bugs.webkit.org/show_bug.cgi?id=268282
rdar://121839192

Reviewed by Simon Fraser.

In 273391@main we added support for scheduling animation resolution for macOS via RemoteLayerTreeEventDispatcher.
On iOS, where we don't use a scrolling thread, we schedule animations through RemoteScrollingCoordinatorProxy
using a dedicated CADisplayLink in a new WKAnimationDisplayLinkHandler. When that display link fires, we resolve
animations on the main thread by calling the new RemoteAcceleratedEffectStack::applyEffectsFromMainThread().

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromMainThread const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(-[WKAnimationDisplayLinkHandler initWithCoordinator:]):
(-[WKAnimationDisplayLinkHandler dealloc]):
(-[WKAnimationDisplayLinkHandler displayLinkFired:]):
(-[WKAnimationDisplayLinkHandler invalidate]):
(-[WKAnimationDisplayLinkHandler schedule]):
(-[WKAnimationDisplayLinkHandler pause]):
(WebKit::RemoteScrollingCoordinatorProxyIOS::~RemoteScrollingCoordinatorProxyIOS):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):

Canonical link: https://commits.webkit.org/274065@main
